File: core/counter/money.py
"""The ticket's arithmetic, in one module, because it is the single most likely
error in the stage.

**Money is stored tax-inclusive.** A Colombian shelf price includes IVA, and the
handoff's own ticket confirms it: `Subtotal $15.600 · Descuento $0 · Total
$15.600`, with no tax line and no tax added to the total. So:

  * `sale_lines.unit_price` is what the customer pays per base unit, IVA
    included;
  * a line's **gross** is `unit_price × quantity` and its **net** is
    `gross − discount`;
  * `tax_amount` is the IVA **contained** in the net, derived from the line's
    own `vat_class` as `net × rate / (100 + rate)`;
  * `sales.subtotal` is the sum of line gross amounts **before** discount,
    `sales.discount` the sum of the line discounts, `sales.total = subtotal −
    discount`, and `sales.tax` the sum of `tax_amount` -- **inside** the total,
    never added to it.

A build that added `tax` to `total` produces a ticket 19% too expensive on
cosmetics and exactly right on medicine, which is the kind of defect a pilot
finds three weeks in. `sales` carries two CHECK constraints saying the same
thing in the database, so the arithmetic is checked on every row this module did
not compute.

**Every figure is a `Decimal` quantized to two places, half-up.** Never a float:
`12345.67` through IEEE 754 and back is not `12345.67`, and this is the one
figure in the product a customer is about to pay.
"""

from decimal import ROUND_HALF_UP, Decimal

from core.models import VAT_RATES

CENTS = Decimal("0.01")
ZERO = Decimal("0.00")

HUNDRED = Decimal("100")


def cents(value) -> Decimal:
    """One money figure, quantized. `None` is zero, because every caller here is
    summing and a null in a sum is a defect two screens away."""
    if value is None:
        return ZERO
    return Decimal(str(value)).quantize(CENTS, rounding=ROUND_HALF_UP)


def contained_tax(net, vat_class: str) -> Decimal:
    """The IVA already inside `net`, for one line.

    `excluded` is not a taxable operation and `exempt` is taxable at 0%; both
    contain nothing, and the difference between them matters to an accountant
    and to S5's document rather than to this sum (§3). An unknown class is
    **not** silently zero -- a class this module does not recognise is a catalog
    the stage has drifted from, and charging no tax on it would be invisible
    until an accountant found it.
    """
    rate = VAT_RATES.get(vat_class)
    if rate is None:
        raise ValueError(
            f"«{vat_class}» no es una clase de IVA reconocida, así que no se "
            "puede saber cuánto impuesto lleva esta línea."
        )
    amount = cents(net)
    if rate == 0 or amount == ZERO:
        return ZERO
    return (amount * rate / (HUNDRED + rate)).quantize(CENTS, rounding=ROUND_HALF_UP)


def line_gross(unit_price, quantity) -> Decimal:
    return cents(cents(unit_price) * Decimal(int(quantity)))


def line_net(unit_price, quantity, discount) -> Decimal:
    return line_gross(unit_price, quantity) - cents(discount)


def line_tax(unit_price, quantity, discount, vat_class) -> Decimal:
    return contained_tax(line_net(unit_price, quantity, discount), vat_class)


class Totals:
    """What a ticket's four figures are, computed from its lines and nothing
    else -- not from what a till sent, which is a browser's arithmetic."""

    __slots__ = ("subtotal", "discount", "tax", "total")

    def __init__(self, subtotal, discount, tax, total):
        self.subtotal = subtotal
        self.discount = discount
        self.tax = tax
        self.total = total

    def __eq__(self, other):
        return isinstance(other, Totals) and (
            self.subtotal,
            self.discount,
            self.tax,
            self.total,
        ) == (other.subtotal, other.discount, other.tax, other.total)

    def __repr__(self):
        return (
            f"Totals(subtotal={self.subtotal}, discount={self.discount}, "
            f"tax={self.tax}, total={self.total})"
        )

    def as_fields(self) -> dict:
        return {
            "subtotal": self.subtotal,
            "discount": self.discount,
            "tax": self.tax,
            "total": self.total,
        }


def totals(lines) -> Totals:
    """Recompute a ticket's four figures from its own lines.

    **The server recomputes rather than trusts.** A till's totals are a
    browser's arithmetic, and the one figure the whole product is measured on is
    not a place to take a client's word -- while the *prices* the till stamped
    are taken exactly as sent, because `sale_lines.unit_price` records what was
    actually charged and no later price list may restate it (§5).

    `lines` is any iterable of objects carrying `unit_price`, `quantity`,
    `discount` and `vat_class`, which is both `SaleLine` and
    `SaleReturnLine` -- a return's money composes exactly as a sale's, because a
    credit note reverses what was charged.
    """
    subtotal = ZERO
    discount = ZERO
    tax = ZERO
    for line in lines:
        subtotal += line_gross(line.unit_price, line.quantity)
        discount += cents(line.discount)
        tax += line_tax(line.unit_price, line.quantity, line.discount, line.vat_class)
    return Totals(subtotal, discount, tax, subtotal - discount)

File: web/src/counter/money.ts
/**
 * The ticket's arithmetic on the till, in **integer centavos**.
 *
 * Never a float. `0.1 + 0.2` is not `0.3`, and this is the one figure in the
 * product a customer is about to pay — a ticket that renders `$15.599,99`
 * because of IEEE 754 is a ticket a cashier argues about at a counter. Every
 * value on the wire is a decimal string, every value in here is an integer
 * number of centavos, and the two conversions are the only place the
 * representation changes. A twelve-digit `numeric(12,2)` is at most 10^14
 * centavos, comfortably inside `Number.MAX_SAFE_INTEGER`.
 *
 * **Money is tax-inclusive**, exactly as it is on the server: `unit_price` is
 * what the customer pays per base unit with IVA in it, `tax_amount` is the IVA
 * *contained* in the line, and a ticket's `total` is `subtotal − descuento`
 * with the tax inside it rather than added to it. A build that added `tax` to
 * `total` produces a ticket 19% too expensive on cosmetics and exactly right on
 * medicine.
 *
 * **The server recomputes all of it on arrival**, so nothing here is the record
 * of anything: this is what the cashier and the customer read while the sale is
 * happening, and `core/counter/money.py` is what the sale *is*. The two agree
 * because they are the same four rules, and the tests hold them to it.
 */

/**
 * The class-to-rate map, mirroring S1's `VAT_RATES`. It is a statute and not a
 * setting — a network does not get to choose it — and a rate changed by decree
 * moves no historical ticket, because `sale_lines.vat_class` and `tax_amount`
 * are both stamped at the moment of sale.
 */
const RATES: Record<string, number> = {
  excluded: 0,
  exempt: 0,
  rate_5: 5,
  rate_19: 19,
};

/** `"15600.00"` → `1560000`. A missing value is zero. */
export function toCents(value: string | number | null | undefined): number {
  if (value === null || value === undefined || value === "") return 0;
  const text = String(value).trim();
  const negative = text.startsWith("-");
  const [whole = "0", fraction = ""] = text.replace("-", "").split(".");
  const cents =
    Number(whole) * 100 + Number((fraction + "00").slice(0, 2) || "0");
  return negative ? -cents : cents;
}

/** `1560000` → `"15600.00"`, which is the shape the wire and the column take. */
export function fromCents(cents: number): string {
  const rounded = Math.round(cents);
  const sign = rounded < 0 ? "-" : "";
  const absolute = Math.abs(rounded);
  return `${sign}${Math.floor(absolute / 100)}.${String(absolute % 100).padStart(2, "0")}`;
}

/** Centavos as pesos, for `@/ui/format`'s `money()` and for nothing else. */
export function pesos(cents: number): number {
  return cents / 100;
}

/** Half-up, matching the server's `ROUND_HALF_UP` — a half-even here and a
 *  half-up there is a one-peso disagreement on every other ticket. */
function round(value: number): number {
  return value < 0 ? -Math.round(-value) : Math.round(value);
}

/** The IVA already inside `net`. An unrecognised class is **not** silently
 *  zero: it is a catalog the till has drifted from. */
export function containedTax(net: number, vatClass: string | null): number {
  const rate = RATES[vatClass ?? ""];
  if (rate === undefined || rate === 0 || net === 0) return 0;
  return round((net * rate) / (100 + rate));
}

export interface Priced {
  quantity: number;
  unit_price: string | number | null;
  discount: string | number | null;
  vat_class: string | null;
}

export function lineGross(line: Priced): number {
  return toCents(line.unit_price) * line.quantity;
}

export function lineNet(line: Priced): number {
  return lineGross(line) - toCents(line.discount);
}

export function lineTax(line: Priced): number {
  return containedTax(lineNet(line), line.vat_class);
}

export interface Totals {
  subtotal: number;
  discount: number;
  tax: number;
  total: number;
  units: number;
}

/** A ticket's four figures, from its own lines and nothing else. */
export function totals(lines: Priced[]): Totals {
  let subtotal = 0;
  let discount = 0;
  let tax = 0;
  let units = 0;
  for (const line of lines) {
    subtotal += lineGross(line);
    discount += toCents(line.discount);
    tax += lineTax(line);
    units += line.quantity;
  }
  return { subtotal, discount, tax, total: subtotal - discount, units };
}
